### ISSUE TITLE
Ideas to Improve GRPO Training Speed

### ISSUE THREAD

**Issue by jp1924** [huggingface/trl#3846]
### Feature request

1. Redundant Old Policy Computation  
   In `GRPOTrainer._generate_and_score_completions`, after receiving completions generated by vLLM, the code concatenates the prompt and completion, then feeds it into the model with `no_grad` to obtain the old policy.  
   My idea is that, since vLLM already generates logits for the concatenated prompt + completion during generation, why not utilize these logits directly? I think this could be beneficial for both colocate and server modes.

   However, a few things need to be checked in vLLM:
   1. Is it possible to obtain the logits for the concatenated prompt + completion?
   2. Is vLLM stable? In my experience, there are sometimes discrepancies between results generated by HuggingFace's `model.generate` and vLLM. I suspect this might be related to recent PRs in vLLM regarding KV cache precision issues, but this needs to be monitored.

2. SGLang Support  
   When using SGLang, which supports Radix-Attention (e.g., in VeRL), training speed reportedly improves by about 1.2x.  
   Besides performance, vLLM still has some instability in generation. Supporting SGLang could provide a good alternative if issues arise with vLLM in the future.

3. Padding-Free Support  
   In both `_generate_and_score_completions` and `compute_loss`, padding tokens are included in the model input. If padding-free functionality is supported, it could help save resources.

### Motivation

While using VeRL, I thought it would be great if TRL also had these features.
Since GRPO is much more compute-intensive compared to standard SFT or DPO, I believe improving these aspects could make training much more resource-efficient.

### Your contribution

Hmm, I'm not sure. Ideally, I’d like to work on all of them, but due to time constraints, I’ll probably have to be selective. If I do contribute, I think I’d most likely work on item 3.

**Comment by LeonEricsson** [comment]
Good suggestions.

1. I'm currently evaluating this, partly due to some of your raised concerns. See #3867 for more details.
2. SGLang is a strong competitor to vLLM so I agree there's value in supporting both.
3. Been thinking about this as well, should be good.
